tests/unit/dataactvalidator/test_b12_award_financial_1.py

```diff
@@ -18,58 +18,52 @@ def test_success(database):
-
-    af = AwardFinancialFactory(ussgl480200_undelivered_or_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
-
-    af = AwardFinancialFactory(ussgl483200_undelivered_or_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
-
-    af = AwardFinancialFactory(ussgl488200_upward_adjustm_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
-
-    af = AwardFinancialFactory(ussgl490200_delivered_orde_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
 
-    af = AwardFinancialFactory(ussgl490800_authority_outl_fyb=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
-
-    af = AwardFinancialFactory(ussgl490800_authority_outl_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
-
-    af = AwardFinancialFactory(ussgl498200_upward_adjustm_cpe=None, by_direct_reimbursable_fun=None)
-    assert number_of_errors(_FILE, database, models=[af]) == 1
+def test_failure(database):
+    """ Test for USSGL 48XX & 49XX (except 487X & 497X) if any one is provided and
+    by_direct_reimbursable_fun is empty the rule fails """
+
+    af_dict = {'by_direct_reimbursable_fun': None, 'ussgl480100_undelivered_or_fyb': None,
+               'ussgl480100_undelivered_or_cpe': None, 'ussgl483100_undelivered_or_cpe': None,
+               'ussgl488100_upward_adjustm_cpe': None, 'ussgl490100_delivered_orde_fyb': None,
+               'ussgl490100_delivered_orde_cpe': None, 'ussgl493100_delivered_orde_cpe': None,
+               'ussgl498100_upward_adjustm_cpe': None, 'ussgl498100_upward_adjustm_cpe': None,
+               'ussgl480200_undelivered_or_fyb': None, 'ussgl480200_undelivered_or_cpe': None,
+               'ussgl483200_undelivered_or_cpe': None, 'ussgl488200_upward_adjustm_cpe': None,
+               'ussgl490200_delivered_orde_cpe': None, 'ussgl490800_authority_outl_fyb': None,
+               'ussgl490800_authority_outl_cpe': None, 'ussgl498200_upward_adjustm_cpe': None}
+
+    keys = ['ussgl480100_undelivered_or_fyb', 'ussgl480100_undelivered_or_cpe', 'ussgl483100_undelivered_or_cpe',
+            'ussgl488100_upward_adjustm_cpe', 'ussgl490100_delivered_orde_fyb', 'ussgl490100_delivered_orde_cpe',
+            'ussgl493100_delivered_orde_cpe', 'ussgl498100_upward_adjustm_cpe', 'ussgl498100_upward_adjustm_cpe',
+            'ussgl480200_undelivered_or_fyb', 'ussgl480200_undelivered_or_cpe', 'ussgl483200_undelivered_or_cpe',
+            'ussgl488200_upward_adjustm_cpe', 'ussgl490200_delivered_orde_cpe', 'ussgl490800_authority_outl_fyb',
+            'ussgl490800_authority_outl_cpe', 'ussgl498200_upward_adjustm_cpe']
+
+    for i in range(len(keys)):
+        af_dict_copy = copy.deepcopy(af_dict)
+        af_dict_copy.pop(keys[i])
```

Comment:
Again, no biggie, but this'd be a little more pythonic:
```python
for key in keys:
    af_dict_copy = copy.deepcopy(af_dict)
    af_dict_copy.pop(key)
```

Also, I don't know how widely used this practice is, but if you want to make a shallow copy (i.e. just another set of keys pointing to the same values), you can do
```python
af_dict_copy = dict(af_dict)
```